frontend/src/components/smart-contracts/TransactionModal.tsx

```diff
@@ -0,0 +1,134 @@
+import React, { useEffect, useState, Dispatch, SetStateAction } from 'react';
+import { ContractTransaction } from 'ethers';
+import { Modal } from 'components/ui/Modal';
+import { ethers } from 'ethers';
+
+export type CallContractCallback = () => Promise<ContractTransaction>;
+
+export type OnConfirmedCallback = (
+  receipt: ethers.providers.TransactionReceipt
+) => void;
+
+export type TransactionError = {
+  message: string;
+};
+
+export enum TXStatus {
+  Ready = 'Ready',
+  Waiting = 'Waiting',
+  Finished = 'Finished',
+  Error = 'Error',
+}
+
+export const TransactionModal = ({
```

Comment:
I think we should be able to use a generic modal and just pass through whatever it needs to display.

Reply:
The TransactionModal is a decorator of Modal.  It handles the complexity of handling the states of a smart contract interaction.  (waiting for the user to interact with their wallet, waiting for blockchain confirmation, and error status).